pytorch_binding/binding.cpp

```diff
@@ -81,42 +82,178 @@ std::tuple<at::Tensor, at::Tensor> rnnt_loss(
+    TORCH_CHECK(xs.dim() == 2, "xs must have 2 dimensions")
+    TORCH_CHECK(xn.size(0) == yn.size(0), "xn and yn shape must be equal (N,)")
+    TORCH_CHECK(ys.numel() == yn.sum().item<int64_t>(),
+                "ys shape must be equal to (sum(yn), )")
+    const at::cuda::OptionalCUDAGuard device_guard(device_of(xs));
+
+    const auto N = xn.size(0);
+    const auto Tm = xn.max().item<int64_t>();     // max of {T_i}
+    const auto Um = yn.max().item<int64_t>() + 1; // max of {U_i}
+    const auto V = xs.size(1);
+
+    auto memPref =
+        (xn * (yn + 1))
+            .cumsum(0, torch::kInt32); // count of frames by current batch
+    auto labelPref = yn.cumsum(0, torch::kInt32); // copy yn
+
+    int64_t STU = memPref[-1].item<int64_t>();
+    TORCH_CHECK(xs.size(0) == STU,
+                "xs shape mismatch with (\\sum{xn*(yn+1)}, )")
+
+    // set begin of memory location of each sequence
+    {
+        auto cumsumMemPref = memPref.index({Slice(0, -1, None)}).clone();
+        auto cumsumLablePref = labelPref.index({Slice(0, -1, None)}).clone();
+        memPref.index_put_({Slice(1, None, None)}, cumsumMemPref);
+        labelPref.index_put_({Slice(1, None, None)}, cumsumLablePref);
+    }
+    memPref[0] = 0;
+    labelPref[0] = 0;
+
+    torch::Tensor gather_xs = torch::empty({STU, 2L}, xs.options());
+    torch::Tensor loc =
+        torch::zeros({STU}, torch::dtype(torch::kInt64).device(xs.device()));
+
+    // gather the labels & blank
+    run_gather_for_compact(
+        xs.data_ptr<float>(), ys.data_ptr<int>(),
+        (unsigned int *)xn.data_ptr<int>(), (unsigned int *)yn.data_ptr<int>(),
+        gather_xs.data_ptr<float>(), loc.data_ptr<long>(),
+        (unsigned int *)memPref.data_ptr<int>(),
+        (unsigned int *)labelPref.data_ptr<int>(), N, Tm, Um, V, blank);
+
+    // the negtive log likelihood
+    torch::Tensor costs = torch::empty({N}, xs.options());
+    //  for maintain the execute status of forward/backward calculation
+    torch::Tensor counts = torch::zeros({ys.numel() * 2 + 2 * N}, ys.options());
+
+    // backward variable of RNN-T
+    torch::Tensor betas = torch::empty({STU}, xs.options());
+
+    torch::Tensor grads, alphas;
+    if (required_grad) {
+        // forward variable of RNN-T
+        alphas = torch::empty_like(betas);
+        grads = torch::empty_like(gather_xs);
+    }
+
+    run_warp_rnnt_compact(
+        (unsigned int *)counts.data_ptr<int>(), alphas.data_ptr<float>(),
+        betas.data_ptr<float>(), gather_xs.data_ptr<float>(),
+        grads.data_ptr<float>(), costs.data_ptr<float>(),
+        (unsigned int *)xn.data_ptr<int>(), (unsigned int *)yn.data_ptr<int>(),
+        (unsigned int *)memPref.data_ptr<int>(),
+        (unsigned int *)labelPref.data_ptr<int>(), N, Tm, Um, fastemit_lambda,
+        required_grad);
+
+    return std::make_tuple(costs, grads, loc);
+}
+
+torch::Tensor rnnt_loss_compact_backward(const torch::Tensor &grad_cost,
```

Comment:
I'm wonder why we need this function separated from `rnnt_loss_compact`? We can invoke this function each time when required_grad==True.

Reply:
Oh, sorry, stupid question. It's propagate the cost gradient, which available only in backward pass.